### PR TITLE
test(grpc): migrate standalone grpc tests to gateway-testing-sdk 

### PR DESCRIPTION
**Pull request by ytvnr**
## Issue

https://gravitee.atlassian.net/browse/APIM-432

### Testing SDK
- add a URLStreamHandler to accept grpc protocol (used to override endpoint port)
- add an AbstractGrpcGatewayTest configuring vertx server (non reactive) and provides utility functions to create service and channel
- clear system properties set by developer thanks to GatewayConfigurationBuilder on server stops. It avoids to keep undesirable properties for following tests
- check existence of endpoints before operating on it (AbstractGatewayTest#updateEndpoints)
- add some utility functions in AbstractGatewayTest and GatewayConfigurationBuilder

### GRPC integration test
- remove the tests from gateway-standalone-container
- migrate them using the gateway testing sdk, with:
  - v3 gateway
  - jupiter gateway, jupiter api
  - jupiter gateway, v3 api

⚠️ [...ner/src/test/java/io/gravitee/gateway/standalone/grpc/GrpcSecuredServerStreamingTest.java](https://github.com/gravitee-io/gravitee-api-management/compare/apim-432-verify-grpc?expand=1#diff-91de28f0b56cffbebe428d3a620be461bd7228f3ad9557c7235f0596ae18269d) was disabled because of race conditions.
I re-enabled it because the gateway testing sdk should handle properly the case:
- system properties are clean on test class stop (top be precise, when vertx stops)
- `@GatewayTest` is a meta-annotation using `@ResourceLock(value = Resources.SYSTEM_PROPERTIES, mode = ResourceAccessMode.READ_WRITE)`. Its purpose is to "lock" the access to a particular resource (here `SYSTEM_PROPERTIES`) preventing to test to modify it at the same time

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-irvhqcbbrn.chromatic.com)

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/apim-432-verify-grpc/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/src/main/java/io/gravitee/apim/gateway/tests/sdk/AbstractGrpcGatewayTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    protected int grpcServerPort;
+
+    protected VertxServer vertxServer;
+    protected ManagedChannel managedChannel;
+
+    @BeforeEach
+    public void setup() {
+        // vertx-grpc needs io.vertx.core.Vertx and do not support the rxjava3 implementation,
+        // so we need a manual instantiation instead of relying on the VertxExtension
+        vertx = Vertx.vertx();
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/src/main/java/io/gravitee/apim/gateway/tests/sdk/AbstractGrpcGatewayTest.java]: I think we can get `io.vertx.core.Vertx` from rxjava3 implementation using the `delegate` getter

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/src/test/java/io/gravitee/apim/gateway/tests/sdk/runner/GatewayRunnerTest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.apim.gateway.tests.sdk.runner;
+
+/**
+ * @author Yann TAVERNIER (yann.tavernier at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+class GatewayRunnerTest {}
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/src/test/java/io/gravitee/apim/gateway/tests/sdk/runner/GatewayRunnerTest.java]: I don't understand why we need this empty class 🤔

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcBidirectionalStreamingIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                            // Adding a latency to simulate multi calls to grpc service
+                            timerID =
+                                vertx.setPeriodic(
+                                    1000,
+                                    periodic -> clientCallStreamObserver.onNext(HelloRequest.newBuilder().setName("You").build())
+                                );
+                        }
+
+                        @Override
+                        public void onNext(HelloReply helloReply) {
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcBidirectionalStreamingIntegrationTest.java:93]: Shouldn't we need to assert the reply?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcSecuredServerStreamingIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    @Override
+                    public void onNext(HelloRequest helloRequest) {
+                        for (int i = 0; i < MESSAGE_COUNT; i++) {
+                            final HelloReply helloReply = HelloReply
+                                .newBuilder()
+                                .setMessage("Hello " + helloRequest.getName() + ", part: " + i)
+                                .build();
+                            responseObserver.onNext(helloReply);
+
+                            responseObserver.onNext(helloReply);
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcSecuredServerStreamingIntegrationTest.java]: Is that normal to call `onNext` twice?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcServerStreamingIntegrationTest.java [hunk 5]
```diff
@@ -13,54 +13,91 @@
...
+        rpcServer.start(
+            event -> {
+                // Get a stub to use for interacting with the remote service
+                StreamingGreeterGrpc.StreamingGreeterStub stub = StreamingGreeterGrpc.newStub(channel);
+
+                // Call the remote service
+                StreamObserver<HelloRequest> requestStreamObserver = stub.sayHelloStreaming(
+                    new StreamObserver<>() {
+                        @Override
+                        public void onNext(HelloReply helloReply) {}
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcServerStreamingIntegrationTest.java]: Shouldn't we assert the reply?

**Review by jgiovaresco — APPROVED** [review]
💪

I'm wondering if the [Vertx Junit extension](https://vertx.io/docs/vertx-junit5/java/) could help us to write better "assertion" in the tests method.
We rely a lot on CountDownLatch, and I'm thinking that checkpoints could help us to analyze what happens in case of a test fails 🤔

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/src/test/java/io/gravitee/apim/gateway/tests/sdk/runner/GatewayRunnerTest.java (hunk 2 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-tests-sdk/src/test/java/io/gravitee/apim/gateway/tests/sdk/runner/GatewayRunnerTest.java]: created by mistake, I delete it :)

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcSecuredServerStreamingIntegrationTest.java (hunk 4 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcSecuredServerStreamingIntegrationTest.java]: no it's an error, I changed it to have assertion on the messages

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcBidirectionalStreamingIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+                            if (testContext.completed()) {
+                                vertx.cancelTimer(timerID);
+                                clientCallStreamObserver.onCompleted();
+                            }
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            fail(throwable.getMessage());
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcBidirectionalStreamingIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcNoEndpointIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            .setName("David")
+            .build();
+
+        // Call the remote service
+        stub.sayHello(
+            request,
+            new StreamObserver<>() {
+                @Override
+                public void onNext(HelloReply helloReply) {
+                    fail("Should not reply");
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcNoEndpointIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcNoEndpointIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                public void onError(Throwable throwable) {
+                    assertThat(throwable).isNotNull().isInstanceOf(StatusRuntimeException.class);
+                    assertThat(((StatusRuntimeException) throwable).getStatus().getCode()).isEqualTo(Status.Code.UNAVAILABLE);
+
+                    testContext.completeNow();
+                }
+
+                @Override
+                public void onCompleted() {
+                    fail("Should not complete");
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcNoEndpointIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcSecuredServerStreamingIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                final StreamObserver<HelloRequest> requestStreamObserver = stub.sayHelloStreaming(
+                    new StreamObserver<>() {
+                        @Override
+                        public void onNext(HelloReply helloReply) {
+                            messageCounter.flag();
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            fail(throwable.getMessage());
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcSecuredServerStreamingIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcServerStreamingIntegrationTest.java [hunk 10]
```diff
@@ -13,54 +13,90 @@
...
+                StreamObserver<HelloRequest> requestStreamObserver = stub.sayHelloStreaming(
+                    new StreamObserver<>() {
+                        @Override
+                        public void onNext(HelloReply helloReply) {
+                            assertThat(helloReply.getMessage()).startsWith("Hello You part");
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            fail(throwable.getMessage());
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcServerStreamingIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnaryRPCIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                        private HelloReply helloReply;
+
+                        @Override
+                        public void onNext(HelloReply helloReply) {
+                            this.helloReply = helloReply;
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            fail(throwable.getMessage());
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnaryRPCIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnaryRPCIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+                            assertThat(helloReply.getMessage()).isEqualTo("Hello You");
+
+                            testContext.completeNow();
+                        }
+                    }
+                );
+            }
+        );
+
+        assertThat(testContext.awaitCompletion(10, TimeUnit.SECONDS)).isTrue();
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnaryRPCIntegrationTest.java]: I'm wondering if this is necessary 🤔

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnknownServiceIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                    assertThat(throwable).isNotNull().isInstanceOf(StatusRuntimeException.class);
+                    final StatusRuntimeException exception = (StatusRuntimeException) throwable;
+
+                    assertThat(exception.getStatus().getCode()).isEqualTo(Status.Code.UNIMPLEMENTED);
+                    assertThat(exception.getMessage()).contains("No context-path matches the request URI.");
+                    testContext.completeNow();
+                }
+
+                @Override
+                public void onCompleted() {}
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnknownServiceIntegrationTest.java]: show we use `textContext.failNow()` here?

File: gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnkownEndpointIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                @Override
+                public void onError(Throwable throwable) {
+                    assertThat(throwable).isNotNull().isInstanceOf(StatusRuntimeException.class);
+                    assertThat(((StatusRuntimeException) throwable).getStatus().getCode()).isEqualTo(Status.Code.UNAVAILABLE);
+
+                    testContext.completeNow();
+                }
+
+                @Override
+                public void onCompleted() {}
```
Comment (jgiovaresco) [inline gravitee-apim-gateway/gravitee-apim-gateway-tests-integration/src/test/java/io/gravitee/gateway/tests/grpc/jupiter/GrpcUnkownEndpointIntegrationTest.java]: show we use `textContext.failNow()` here?

**Review by jgiovaresco — APPROVED** [review]

**Pull request merged**